### ISSUE TITLE
xgboost.spark generates subprocesses bypassing cluster's resource monitoring

### ISSUE THREAD

**Issue by TurtlelovePig** [dmlc/xgboost#11326]
Environment: xgboost 2.1.1, pyspark 3.4.1, yarn
Symptoms: During the training of an XGBoost model using the xgboost.spark, multiple subprocesses are spawned by XGBoost. These subprocesses bypass the cluster's resource monitoring (YARN) and request additional memory outside the Spark-managed quota. When the cluster is under high load, this uncontrolled memory allocation leads to resource exhaustion and eventual cluster crashes 
Config & code:
spark = SparkSession.builder \
    .appName("PySparkLightGBM_SpillTest") \
    .config("spark.local.dir", "/work_fast/tmp/") \
    .config("spark.port.maxRetries","1000") \
    .config("spark.speculation", "true") \
    .config("spark.yarn.executor.memoryOverhead", "30000") \
    .config("spark.network.timeout", "2400s")\
    .config("spark.task.maxFailures", "20") \
    .config("spark.executor.heartbeatInterval", "1200s") \
    .config("spark.driver.maxResultSize","60g") \
    .config("spark.driver.cores","15")\
    .config("spark.executor.instances", "20") \
    .config("spark.executor.cores", "20") \
    .config("spark.executor.memory", "80g") \
    .config("spark.executor.memoryOverhead", "20g") \
    .config("spark.memory.offHeap.size", "30g") \
    .config("spark.driver.memory", "60g") \
    .config("spark.sql.shuffle.partitions", "1200") \
    .config("spark.default.parallelism", "1000") \
    .config("spark.serializer", "org.apache.spark.serializer.KryoSerializer") \
    .config("spark.kryoserializer.buffer.max", "1024m") \
    .config("spark.memory.fraction", "0.7") \
    .config("spark.memory.storageFraction", "0.8") \
    .config("spark.shuffle.spill.compress", "true") \
    .config("spark.shuffle.compress", "true") \
    .config("spark.memory.offHeap.enabled", "true") \
    .config("spark.sql.broadcastTimeout", "1200")\
   .config("spark.shuffle.io.maxRetries", "30") \
   .config("spark.task.cpus", 3) \
   .config("spark.shuffle.io.retryWait", "600s") \
   .config("spark.reducer.maxSizeInFlight", "300m") \
   .config("spark.shuffle.file.buffer", "10m") \
   .config("spark.pyspark.python", f"{driver_python_path}") \
   .config("spark.executorENV.PYTHONPATH", "/workspace/home/anaconda3/xgb/lib310/python3.10/site-packages") \
   .config("spark.driverENV.PYTHONPATH", "/workspace/home/anaconda3/xgb310/lib/python3.10/site-packages") \
    .config("spark.executor.extraJavaOptions","-XX:+UseG1GC -XX:InitiatingHeapOccupancyPercent=35")\

        xgb = SparkXGBRegressor(num_workers=120,n_estimators=iteration_number,callbacks=[lr_scheduler],force_repartition=True,**param_set)
        model = xgb.fit(train_data)

thanks for your help!

**Comment by TurtlelovePig** [comment]
![Image](https://github.com/user-attachments/assets/22ce7e88-a7ed-46ea-ad1f-a128f5e305ad)

**Comment by trivialfis** [comment]
The pyspark module is defined here https://github.com/dmlc/xgboost/blob/master/python-package/xgboost/spark/__init__.py I don't think there's any code that can fork or use `subprocess`.

**Comment by TurtlelovePig** [comment]
> The pyspark module is defined here https://github.com/dmlc/xgboost/blob/master/python-package/xgboost/spark/__init__.py I don't think there's any code that can fork or use `subprocess`.

but it does generate some "process" and consume lots of memory that not monitored by yarn

**Comment by trivialfis** [comment]
cc @wbo4958 might have better insight into process management by spark.

Overall, Spark manages the Python process. Data is transferred from the Java process to the Python process using Arrow, which Pyspark also manages. XGBoost takes the data [here](https://github.com/dmlc/xgboost/blob/9ab7567de4c6a197756c4726ee63966e38f33232/python-package/xgboost/spark/data.py#L259) without launching any new process (but does launch a lot of new threads).

Perhaps Spark doesn't track the Python memory usage yet and can only deal with Java processes?

**Comment by trivialfis** [comment]
After some googling, the Python memory usage seems to be part of the "overheadMemory" for yarn instead of being properly tracked.

**Comment by wbo4958** [comment]
Hi @TurtlelovePig, spark default to use Daemon to create the python workers which will run the actual python code. see https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala#L123

So the multiple python processes are expected. You could disable daemon by setting `spark.python.use.daemon=false`

**Comment by wbo4958** [comment]
You set `spark.executor.cores=20` and `spark.task.cpus=3` which results in 6 tasks run on the same executors at the same time, and you set `spark.executor.memory=80G`  and `spark.executor.memoryOverhead=20G` which means each task will take 16.7G on average. What's the data size?

**Comment by TurtlelovePig** [comment]
> You set `spark.executor.cores=20` and `spark.task.cpus=3` which results in 6 tasks run on the same executors at the same time, and you set `spark.executor.memory=80G` and `spark.executor.memoryOverhead=20G` which means each task will take 16.7G on average. What's the data size?

sorry for late reply, the total datasize is about 2T

**Comment by TurtlelovePig** [comment]
> Hi [@TurtlelovePig](https://github.com/TurtlelovePig), spark default to use Daemon to create the python workers which will run the actual python code. see https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala#L123
> 
> So the multiple python processes are expected. You could disable daemon by setting `spark.python.use.daemon=false`

thank you, so what will happen if i setting this to 'false', will slow down the train obviously?

**Comment by trivialfis** [comment]
You will have to experiment. We haven't tested these configurations from spark. Feel free to share your results. ;-)

**Comment by wbo4958** [comment]
Hi @TurtlelovePig, Since the datasize is 2T, so each task should train like 2T/120 = 17G which causes OOM which is expected.

Typically, you could tune this issue by decreasing the dataset size.

**Comment by TurtlelovePig** [comment]
@wbo4958 @trivialfis I have tried to train with a much smaller dataset, but it still spawn multiple python processes bypassing the yarn's monitor; additionally, i tried to set spark.python.use.daemon=false, and it will raise an timeout error; so i wonder if this is caused by python, can i use xgboost4j instead on yarn to avoid the multiple processes?

**Comment by wbo4958** [comment]
@TurtlelovePig, The multi-python processes are expected. and I don't think they're going to occupy much memory. xgboost4j  could work without python process.

But as for this issue, I think you could set 

`spark.task.cpus=20` and num_workers=20 to test if they're going to work in a smaller dataset, then you could increase num_workers

**Comment by wbo4958** [comment]
@TurtlelovePig, or you can try below configurations on small dataset. Since xgboost training is using non-heap memory instead of heap memory.

``` console
.config("spark.executor.memory", "20g")
.config("spark.executor.memoryOverhead", "80g")
```

**Comment by TurtlelovePig** [comment]
> [@TurtlelovePig](https://github.com/TurtlelovePig), or you can try below configurations on small dataset. Since xgboost training is using non-heap memory instead of heap memory.
> 
> .config("spark.executor.memory", "20g")
> .config("spark.executor.memoryOverhead", "80g")

thank u. i will try it now

**Comment by wbo4958** [comment]
BTW, @TurtlelovePig, I don't think you need `spark.memory.offHeap.size` which is for spark internal memory management.

You could change to

.config("spark.executor.memory", "30g")
.config("spark.executor.memoryOverhead", "100g")

**Comment by TurtlelovePig** [comment]
@wbo4958 I have tried the config on a 60g size dataset
.config("spark.executor.memory", "20g")
.config("spark.executor.memoryOverhead", "80g")
but it will raise the following error

![Image](https://github.com/user-attachments/assets/00f90a4b-1bc3-4628-add5-fe7c348ed05a)

**Comment by TurtlelovePig** [comment]
> [@TurtlelovePig](https://github.com/TurtlelovePig), or you can try below configurations on small dataset. Since xgboost training is using non-heap memory instead of heap memory.
> 
> .config("spark.executor.memory", "20g")
> .config("spark.executor.memoryOverhead", "80g")

@wbo4958 I tried this config on a 75g size dataset, and there are still some python process, the following picture display the processes on one of the 7 nodes;

![Image](https://github.com/user-attachments/assets/bc0a91fc-67ed-4ab4-9611-d59f639054c4)

Additionally, i monitor the memory usage on yarn, as the picture following shows

![Image](https://github.com/user-attachments/assets/d5e40d75-fd92-48c5-8f78-26b5b021032f)

**Comment by wbo4958** [comment]
@TurtlelovePig, Move your eyes from the number of the python processes. Just to ensure if the configurations can make the xgboost training work.

**Comment by trivialfis** [comment]
Considering the new processes are launched by Spark instead of XGBoost, I will close this issue now. If there are other questions, feel free to open new ones.

> [@TurtlelovePig](https://github.com/TurtlelovePig), Move your eyes from the number of the python processes. Just to ensure if the configurations can make the xgboost training work.

The new processes are necessary for PySpark to work and are harmless.

**Comment by TurtlelovePig** [comment]
@trivialfis @wbo4958 After some attempts, the problem was found to be that the config ".config("spark.executor.memoryOverhead", "100g")" doesn't work. I mean whatever this config i set, such as "1g" or "100g", the real memory consumed during train is determined by the size of dataset

**Comment by wbo4958** [comment]
Yes, that's true. `spark.executor.memoryOverhead` is for spark internal native operations in the java process. While the xgboost training is running on a separate worker of a python worker, so  if you didn't set pyspark.memory, xgboost training will use all the memory of the yarn container by default.

**Comment by TurtlelovePig** [comment]
> Yes, that's true. `spark.executor.memoryOverhead` is for spark internal native operations in the java process. While the xgboost training is running on a separate worker of a python worker, so if you didn't set pyspark.memory, xgboost training will use all the memory of the yarn container by default.

thanks; another question is that, if dataset size is X, and i use SparkXGBoostRegressor with 'hist' method, is the memory actually consumed 4*X? or something else
